src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java

```diff
@@ -811,8 +813,51 @@ private static class ImageHelper implements ImageProvider {
                     Map<String, Path> modsPaths,
                     ByteOrder order,
                     Path packagedModulesPath,
-                    boolean ignoreSigning) throws IOException {
-            this.order = order;
+                    boolean ignoreSigning,
+                    boolean verbose,
+                    PrintWriter log) throws IOException {
+            if (order != null) {
+                this.order = order;
+            } else {
+                Path javaBasePath = modsPaths.get("java.base");
+                assert javaBasePath != null : "java.base module path is missing";
+                if (this.isJavaBaseFromCurrentPlatform(javaBasePath)) {
+                    // this implies that the java.base module used for the target image
+                    // will correspond to the current platform. So this isn't an attempt to
+                    // build a cross-platform image. We use the current platform's endianness
+                    // in this case
+                    this.order = ByteOrder.nativeOrder();
+                } else {
+                    // this is an attempt to build a cross-platform image. We now attempt to
+                    // find the target platform's arch and thus its endianness from the java.base
+                    // module's ModuleTarget attribute
+                    Optional<ResolvedModule> javaBase = cf.findModule("java.base");
```

Comment:
`ModuleTarget` is read by `DefaultImageReader::storeFiles` as well.   Perhaps it should be refactored so that the target `Platform` can be passed to `DefaultImageBuilder`.

Reply:
I've updated the PR to pass along this already parsed target platform to the `DefaultImageBuilder`. This code is new to me and I think I made sure (by checking references and the package names) that I only updated the internal classes without impacting any user facing APIs. If however, this impacts any user facing API, please do let me know.